### PR TITLE
Reducing RAM Footprint

### PR DESCRIPTION
**Pull request by Felalolf**
This is not a real PR. I want to run all checks on a trivial modification to identify if there is a problem with the CI.

### CODE REVIEW

**Comment by ArquintL** [comment]
The 137 exit code is caused by Gobra's huge RAM usage. It stays a miracle why we have not observed this behavior earlier. Rerunning the CI for past commits (for which it the CI succeeded back then) has now a high likelihood of failing too.

I've identified a memory issue with the `GobraExecutionContext`: We create a new one for every test case but we did not terminate it. I've added a similar termination thingy to ViperServer but apparently forgot to backport it to Gobra. I was able to observe a memory reducing from ~5.5GB to ~3GB when executing the tests in `gobra-test.jar` (directly using java instead of sbt) after fixing it.
Running the tests with `sbt test` however leads to a higher memory consumption: `pidstat` reports two instead of a single `java` process and both of them are using roughly 3GB of RAM. This together is very close to the maximum RAM amount available on Ubuntu and Windows runners provided by GitHub. I have not figured out what causes this overhead when using `sbt`.

To keep track of the memory issues in the future, `pidstat` is run in parallel while executing the tests and the maximum RAM amount of z3 and java get calculated. Afterwards, corresponding annotations are created such that memory usage becomes clearly visible on GitHub. Note however that the two Java processes when running `sbt test` are not correctly taken into account as only the maximum instead of their sum is considered.

I propose to merge this improvements for now such that development is no longer stalled. However, it's also clear that this PR does not address all issues yet.

**Comment by ArquintL** [comment]
This PR should be squash committed

**Comment by ArquintL** [comment]
+reviewer @Felalolf

File: src/test/scala/viper/gobra/GobraTests.scala [hunk 1]
```diff
@@ -57,6 +59,12 @@ class GobraTests extends AbstractGobraTests with BeforeAndAfterAll {
 
         val executor: GobraExecutionContext = new DefaultGobraExecutionContext()
         val (result, elapsedMilis) = time(() => Await.result(gobraInstance.verify(config)(executor), Duration.Inf))
+        val startTime = System.currentTimeMillis()
+        // terminate executor with a larger timeout such that we can distinguish a timeout from terminate taking quite long
+        executor.terminate(10 * execution_context_terminate_timeout_ms)
+        val terminateDurationMs = System.currentTimeMillis() - startTime
+        // check whether timeout has been exceeded and fail test accordingly:
+        assert(terminateDurationMs < execution_context_terminate_timeout_ms)
```
Comment (jcp19) [inline src/test/scala/viper/gobra/GobraTests.scala:67]: This code seems to be repeated in various places, could we abstract it in a function?

**Review by jcp19 — APPROVED** [review]
Looks good to me. Nonetheless, I found some places where the code can be simplified.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -83,6 +83,98 @@ jobs:
...
+        # we transform bash variables into awk variables to not cause problems with bash's variable substitution
+        # the required and optional JSON fields can be found in section "Properties of the `annotations` items" here:
+        # https://docs.github.com/en/rest/reference/checks#create-a-check-run--parameters
+        # after computing the memory usage (in KB) a few more computations are performed. At the very end, all (local)
+        # environment variables are exported to `$GITHUB_ENV` such that they will be available in the next step as well.
+        run: |
+          MAX_JAVA_KB="$(awk -v max=0 '/java$/ {if($8>max){max=$8}}END{print max}' pidstat.txt)"
+          MAX_Z3_KB="$(awk -v max=0 '/z3$/ {if($8>max){max=$8}}END{print max}' pidstat.txt)"
+          MAX_JAVA_GB=$(awk -v MAX_JAVA_KB=$MAX_JAVA_KB 'BEGIN {print MAX_JAVA_KB / 1000 / 1000}')
+          MAX_Z3_GB=$(awk -v MAX_Z3_KB=$MAX_Z3_KB 'BEGIN {print MAX_Z3_KB / 1000 / 1000}')
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply .github/workflows/test.yml:134]: Wouldn't this perform an int division? I'm using `awk` to get floats

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -83,6 +83,98 @@ jobs:
...
+          JAVA_LEVEL=$(awk -v MAX_JAVA_GB=$MAX_JAVA_GB 'BEGIN { print (MAX_JAVA_GB>${{ env.JAVA_FAILURE_THRESHOLD_GB }}) ? "failure" : (MAX_JAVA_GB>${{ env.JAVA_WARNING_THRESHOLD_GB }}) ? "warning" : "notice"}')
+          Z3_LEVEL=$(awk -v MAX_Z3_GB=$MAX_Z3_GB 'BEGIN { print (MAX_Z3_GB>${{ env.Z3_FAILURE_THRESHOLD_GB }}) ? "failure" : (MAX_Z3_GB>${{ env.Z3_WARNING_THRESHOLD_GB }}) ? "warning" : "notice"}')
+          CONCLUSION=$(awk -v JAVA_LEVEL=$JAVA_LEVEL -v Z3_LEVEL=$Z3_LEVEL 'BEGIN { print (JAVA_LEVEL=="failure" || Z3_LEVEL=="failure") ? "failure" : "success"}')
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply .github/workflows/test.yml:137]: Same here: I don't think bash can perform float comparison

File: .github/workflows/test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jcp19) [reply .github/workflows/test.yml:134]: Oh, you're right, my bad!

**Comment by ArquintL** [comment]
I've now improved reporting of memory usage by taking the sum when there are multiple instances of the same process. To do so, I consider the timestamp reported by `pidstat` which might not be perfect as it is not guaranteed that all entries of a report have the same timestamp but it should result in a rough estimate in practice

**Comment by ArquintL** [comment]
In addition to our own memory issue, a change in GitHub's virtual environment has actually triggered the issues we have observed. [more info](https://github.com/actions/virtual-environments/issues/3773)

**Pull request merged**